source/MRViewer/MRSubfeaturePickerWidget.h

```diff
@@ -0,0 +1,218 @@
+#pragma once
+
+#include "MRMesh/MRFeatureObject.h"
+#include "MRMesh/MRFeatures.h"
+#include "MRViewer/MRObjectHighlighter.h"
+#include "MRViewer/MRViewerEventsListener.h"
+#include "MRViewer/MRViewport.h"
+#include "MRViewer/exports.h"
+
+#include <span>
+
+namespace MR
+{
+
+// A helper class to pick into subfeatures.
+// This is the minimal GUI-less version. Use `SubfeaturePickerWidget` to also get GUI and object highlighting.
+// Creates temporary feature objects for the subfeatures and lets you pick into them.
+// The workflow is:
+// 1. Construct this as a member variable in your plugin.
+// 2. Call `updateTemporaryObjects()` in `preDraw_()`.
+// 3. Call `removeTemporaryObjects()` in `onDisable_()`.
+// 4. Call `pickObject()` when you need it. If you do it from `mouseMove_()`, consult the comments on `pickObject()`.
+//    Alternatively, you can use `temporarilyAdjustSubfeaturesForPicking()` to run your own customized picker.
+//    You can do anything with the objects you receive from the picker. Subfeatures are invisible by default,
+//      but you can make the picked one visible manually.
+//    You can analyze the picked subfeatures using `findSubfeature()`.
+class SubfeaturePickerWidgetLow
+{
+public:
+    using ObjectPred = std::function<bool( const FeatureObject& feature, bool isAncillary )>;
+
+    // If set, this is called for every new subfeature object we create.
+    std::function<void( FeatureObject& )> initializeSubfeature;
+
+    // Ignore featues that have `Subfeatures` visual property disabled.
+    bool respectSubfeatureVisualProperty = true;
+    // Whether ancillary features should be processed.
+    bool allowAncillaryObjects = false;
+
+    MRVIEWER_API static float defaultInfiniteExtent();
+    // How large we should make "infinite" features.
+    std::function<float()> getInfiniteExtent = defaultInfiniteExtent;
+
+    // Temporarily override feature sizes to those values for picking:
+    float pickerPointSize = 16;
+    float pickerLineWidth = 10;
+
+    struct SubfeatureData
+    {
+        std::string name;
+        Features::Primitives::Variant subfeature;
+        std::shared_ptr<FeatureObject> object;
+    };
+
+    struct ObjectData
+    {
+        Features::Primitives::Variant feature;
+        std::vector<SubfeatureData> subfeatures;
+    };
+
+    using ObjectMap = std::map<std::shared_ptr<Object>, std::shared_ptr<ObjectData>>;
+    // Returns all tracked feature objects.
+    [[nodiscard]] MRVIEWER_API const ObjectMap& getObjects() const;
+
+    // Update the temporary objects representing subfeatures.
+    MRVIEWER_API void updateTemporaryObjects( ObjectPred objectPredicate = nullptr );
```

Comment:
what are "temporary objects"?

Reply:
One temporary object per subfeature, we need them for the picker and to be able to highlight them.